### PR TITLE
Allow arrow functions in JSX props

### PR DESCRIPTION
**Pull request by gilbox**
Rationale: 

- Improves code readability. 
- Results in simpler solutions than the alternative.

### CODE REVIEW

**Comment by kesne** [comment]
You should reference the eslint rule (which in this case is also `react/jsx-no-bind`).

**Comment by tleunen** [comment]
:-1: 
You can always find a way to avoid this. For example passing the item and the index as props, with the onClick callback.

**Comment by ljharb** [comment]
@tleunen what about at the bottom level, where you directly create a native DOM element?

**Comment by tleunen** [comment]
What do you mean by that?

To me, it's way better to have
```jsx
function ItemList(props) {
  return (
    <ul>
      {props.items.map((item, index) => (
        <Item
          key={item.key}
          item={item}
          listIndex={index}
          onItemClick={doSomethingWith}
        />
      ))}
    </ul>
  );
}
```
And then Item can attach the `onClick` event where needed and call `props.doSomethingWith(props.item.name, props.listIndex)`

The downside is that it has to know the list index, but well.. I think it's better than creating a new bound function every time.

**Comment by ljharb** [comment]
@tleunen ok so what does the implementation of `Item` look like? *somewhere* you are creating the `<a>` or `<input>` and that place needs to create a new function for the event handler.

**Comment by tleunen** [comment]
Sure, but you only create it once

```jsx
class Item extends React.Component {
    constructor(props) {
        super(props);
        this.handleClick = this.handleClick.bind(this);
    }
    handleClick(e) {
        this.props.doSomethingWith(props.item.name, props.listIndex);
    }
    render() {
        return <button onClick={this.handleClick}>Click me</button>
    }
}

**Comment by ljharb** [comment]
Right, so the question is really whether `Item` should be an SFC with an arrow function in `onClick`, or whether it should be a `class`-based component like in your example.

**Comment by kesne** [comment]
To me, the arrow function seems a lot cleaner: 

```javascript
function Item({ doSomethingWith, item, listIndex }) {
  return <button onClick={() => doSomethingWith(item.name, listIndex)}>Click me</button>;
}
```

**Comment by tleunen** [comment]
I know React can apply some performance optimizations on the SFC, but would that be better than having the class without recreating the function at every render?

**Comment by ljharb** [comment]
First, as I have recently learned, the only advantage with the constructor-binding is that `.bind` is slow - creating new functions is not.
Second, performance should always be a secondary concern to code clarity and readability - something would have to be *very* slow to trump readability concerns, and I've been convinced that creating non-bound functions as event handlers (not as props, however, just as bottom-level event handlers) is not a performance hit due to the design of React.

Thus this is mostly a question of subjective readability, I think.

**Comment by tleunen** [comment]
Make sense. So creating non-bound function is fine, but not otherwise.

**Comment by lelandrichardson** [comment]
I am in favor of allowing arrow functions in render methods. We should continue disallowing `.bind(this)` though

File: packages/eslint-config-airbnb/rules/react.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ module.exports = {
     // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md
     'react/jsx-no-bind': [2, {
       'ignoreRefs': false,
```
Comment (lelandrichardson) [inline packages/eslint-config-airbnb/rules/react.js]: will we be able to do this, then: `ref={foo => this.foo = foo}`?

Comment (ljharb) [inline packages/eslint-config-airbnb/rules/react.js]: only if this was set to `true`, as i understand it

Comment (lelandrichardson) [inline packages/eslint-config-airbnb/rules/react.js]: if arrow fns were allows, seems like we should change that as well?

**Comment by tleunen** [comment]
@lelandrichardson Would `this` be allowed in an arrow function?
I know Babel doesn't create a bound function, but how does the browser see the arrow function with the usage of `this` inside?

**Comment by ljharb** [comment]
@tleunen the same as one without. it basically does a `var self = this` and creates a closure.

File: packages/eslint-config-airbnb/rules/react.js (hunk 1 above)
Comment (kesne) [inline packages/eslint-config-airbnb/rules/react.js]: I agree, arrows for refs are useful when you want to use inputs in a SFC.

**Comment by gilbox** [comment]
Which is more readable?

*(A)* Foo passess the data down and Bar passes it back up again
```
function Foo({ data }) {
  return (
    <Bar onSomething={Actions.zoom} data={data} />
  )
}
```

*(B)* Use a closure
```
function Foo({ data }) {
  return (
    <Bar onSomething={() => Actions.zoom(data)} />
  )
}
```

The answer is *(B)* because you don't need to look at the code of `Bar` to understand what data is being passed into `Actions.zoom`. This is why we should usually favor using closures even for non-leaf nodes. Passing the same data down and then back up again is an anti-pattern imo.

**Comment by ljharb** [comment]
I agree that B is more readable - however with the rules as-is on master, you'd model that as a stateful component.

**Pull request merged**